glslang/Public/ShaderLang.h

```diff
@@ -292,25 +292,84 @@ class TShader {
     void setPreamble(const char* s) { preamble = s; }
 
     // Interface to #include handlers.
+    //
+    // To support #include, a client of Glslang does the following:
+    // 1. Call setStringsWithNames to set the source strings and associated
+    //    names.  For example, the names could be the names of the files
+    //    containing the shader sources.
+    // 2. Call parse with an Includer.
+    //
+    // When the Glslang parser encounters an #include directive, it calls
+    // the Includer's include method with the the requested include name
+    // together with the current string name.  The returned IncludeResult
+    // contains the fully resolved name of the included source, together
+    // with the source text that should replace the #include directive
+    // in the source stream.  After parsing that source, Glslang will
+    // release the IncludeResult object.
     class Includer {
     public:
-        // On success, returns the full path and content of the file with the given
-        // filename that replaces "#include filename". On failure, returns an empty
-        // string and an error message.
-        virtual std::pair<std::string, std::string> include(const char* filename) const = 0;
+        typedef enum {
+          EIncludeRelative, // For #include "something"
+          EIncludeStandard  // Reserved. For #include <something>
+        } IncludeType;
+
+        // An IncludeResult contains the resolved name and content of a source
+        // inclusion.
+        struct IncludeResult {
+          // For a successful inclusion, the fully resolved name of the requested
+          // include.  For example, in a filesystem-based includer, full resolution
+          // should convert a relative path name into an absolute path name.
+          // For a failed inclusion, this is an empty string.
+          std::string file_name;
+          // The content and byte length of the requested inclusion.  The
+          // Includer producing this IncludeResult retains ownership of the
+          // storage.
+          // For a failed inclusion, the file_data
+          // field points to a string containing error details.
+          const char* file_data;
+          const size_t file_length;
+          // Include resolver's context.
+          void* user_data;
+        };
+
+        // Resolves an inclusion request by name, type, and current source name.
```

Comment:
Need to mention the inclusion depth.

"Resolves an inclusion request by name, type, current source name, and inclusion depth."